sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -579,36 +582,60 @@ static inline int find_codec_info_acpi(const u8 *acpi_id)
 	return i;
 }
 
-static int get_sdw_dai_link_num(struct snd_soc_acpi_mach_params *mach_params)
+/*
+ * get BE dailink number and CPU DAI number based on sdw link adr.
+ * Since some sdw slaves may be aggregated, the CPU DAI number
+ * may be larger than the number of BE dailinks.
+ */
+static int get_sdw_dailink_info(const struct snd_soc_acpi_link_adr *links,
+				int *sdw_be_num, int *sdw_cpu_dai_num)
 {
 	const struct snd_soc_acpi_link_adr *link;
-	int num = 0;
+	bool group_visited[SDW_MAX_GROUPS];
+	int cpu_dai_num = 0;
+	int be_num = 0;
+	int i, stream;
 
-	link = mach_params->links;
-	if (!link)
+	if (!links)
 		return -EINVAL;
 
-	for (; link->num_adr; link++) {
-		int part_id, id;
-		u64 adr;
+	for (i = 0; i < SDW_MAX_GROUPS; i++)
+		group_visited[i] = false;
 
-		adr = link->adr_d[0].adr;
+	for (link = links; link->num_adr; link++) {
+		const struct snd_soc_acpi_endpoint *endpoint;
+		int part_id, codec_index;
+		u64 adr;
 
+		adr = link->adr_d->adr;
 		part_id = SDW_PART_ID(adr);
-		id = find_codec_info_part(part_id);
-		if (id < 0)
-			return 0;
+		codec_index = find_codec_info_part(part_id);
+		if (codec_index < 0)
+			return codec_index;
+
+		endpoint = link->adr_d->endpoints;
 
-		/* support playback */
-		if (codec_info_list[id].direction[0])
-			num++;
+		/* count DAI number for playback and capture */
+		for_each_pcm_streams(stream) {
+			if (!codec_info_list[codec_index].direction[stream])
+				continue;
 
-		/* support capture */
-		if (codec_info_list[id].direction[1])
-			num++;
+			cpu_dai_num++;
+
+			/* count BE for each non-aggregated slave or group */
+			if (!endpoint->aggregated ||
+			    !group_visited[endpoint->group_id])
+				be_num++;
+		}
+
+		if (endpoint->aggregated)
+			group_visited[endpoint->group_id] = true;
 	}
 
-	return num;
+	*sdw_be_num = be_num;
+	*sdw_cpu_dai_num = cpu_dai_num;
```

Comment:
can we remove the temp variables and use the sdw_be_num and sdw_cpu_dai_num directly?

Reply:
yes, done!